### ISSUE TITLE
Reproduce the example benchmarks

### ISSUE THREAD

**Issue by jimdowling** [uber/horovod#21]
Will you be making the benchmark programs available in the examples section (Resnet, Inception, VGGNet)? That would enable us to reproduce your numbers. Thanks.

**Comment by alsrgv** [comment]
Hi, @jimdowling.

We have made [a fork](https://github.com/alsrgv/benchmarks/tree/horovod) of official TF benchmarks available.

In that fork, we replaced all the regular multi-tower / multi-node code with Horovod in [this](https://github.com/alsrgv/benchmarks/commit/86bf2f9269dbefb4e57a8b66ed260c8fab84d6c7) commit.

## Our test setup

1. We install Horovod following the [Have GPUs?](https://github.com/uber/horovod#have-gpus) guide.

2. We run our tests with regular `mpirun`:
```
$ NCCL_DEBUG=INFO mpirun \
  -mca btl_openib_receive_queues P,128,32:P,2048,32:P,12288,32:P,131072,32 \
  -x LD_LIBRARY_PATH \
  -x NCCL_DEBUG \
  -n 16 \
  -H host1:4,host2:4,host3:4,host4:4 \
  python tf_cnn_benchmarks.py --model resnet101 --batch_size 64 --horovod_device gpu
```

3. We then look for log lines "Total images/sec" from worker 0 and take median value across multiple runs.

Please share the #s that you will get on your hardware.

**Comment by jimdowling** [comment]
That's brilliant @alsrgv . Two Deeplearning11's on their way with 40Gb infiniband. Will report back.

**Comment by alsrgv** [comment]
Great, looking forward to it!

With so many GPUs on the PCIx bus and a shared set of RAM channels, it's important to make sure you're not saturating them. Would be interesting to see 1x10 GPU vs 2x5 GPU comparison.

All the benchmarks we ran so far were on 4 GPU PCIx servers.

**Comment by jimdowling** [comment]
These are 1x10 SingleRoot supermicro machines. 2 weeks delivery, we've been told, so end of the month, i guess.

**Comment by alsrgv** [comment]
We can leave this issue open till you get the numbers.

**Comment by k9sret** [comment]
Are you training alexnet with real data?

**Comment by alsrgv** [comment]
@jimdowling, any luck with your tests?

**Comment by jimdowling** [comment]
@alsrgv  - just got delivery of the 1st server, waiting on the 2nd. Long wait time on the supermicro 10GPU machines. Will take another couple of weeks, i guess.

**Comment by hfutxrg** [comment]
Hi @alsrgv , 

I have tried horovod with resnet50 and ILSVRC2012 dataset. Now the speedup is ~1.4x for two nodes. I think the performance can be further improved with setting like "-mca btl_openib_receive_queues P,128,32:P,2048,32:P,12288,32:P,131072,32". But your P values settings does not work for me. Could you explain a bit more about what this setting means?  I searched online and found it's something related to IB routing. But I still don't know how to set it for our own system. Could you give some guide on how to set these values? 

By the way, do you have any technical document about the strategy of your parallelization? The files in doc folder is still not very clear to me. I want to know more details. 

Thanks very much!
Rengan

**Comment by alsrgv** [comment]
Hi @hfutxrg,

Can you clarify your setup:
1. How many servers, and how many GPU cards do you run on?
2. What network cards do you have?
3. What command did you run to install Horovod?  Did you specify `HOROVOD_GPU_ALLREDUCE=NCCL` if you have GPUs?
4. What is your `mpirun` command?

I just ran resnet50 benchmark on our 16 GPU (4x4) cluster, and I got ~14.3x.

We're writing a blog post about Horovod which will have more details, it will be available in a few days.

**Comment by hfutxrg** [comment]
@alsrgv 14.3x speedup is impressive. I hope I also can get that speedup.

Here are my setup:
1. I have up to 4 servers and each server has 4 P100 GPUs. Now first I am trying only 2 servers and then scale to 4 servers later.
2. My network card is Mellanox EDR (100Gb/s).
3. I used the same command on the docs page:
```HOROVOD_NCCL_HOME=/cm/shared/apps/nccl/2.0.4 HOROVOD_GPU_ALLREDUCE=NCCL HOROVOD_GPU_ALLGATHER=MPI HOROVOD_GPU_BROADCAST=MPI pip install --no-cache-dir horovod```

    My MPI version is 2.1.1 and was compiled with cuda 9.

4. My mpirun command is as follows:
```export NCCL_DEBUG=INFO```
```mpirun --allow-run-as-root \```
   ``` -n 8 \```
   ``` -npernode 4 \```
    ```python tf_cnn_benchmarks.py \```
    ```--model=resnet50 \```
    ```--batch_size=64 \```
    ```--horovod_device=gpu \```
    ```--num_batches=2000 \```
    ```--data_dir=${DATASET_DIR} \```
   ``` --data_name="imagenet" \```
   ``` --print_training_accuracy ```

  My mpirun command is in a slurm script and I make 4 processes run on one node and another 4 processes run on another node.

In my output file, I also noticed there are many following messages:
node003:8342:8486 [0] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported
node003:8342:8486 [0] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported
node003:8342:8486 [0] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported
node003:8342:8486 [0] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported
node003:8343:8484 [1] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported
node003:8343:8484 [1] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported
......
node014:36616:36771 [0] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported
node014:36616:36771 [0] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported
node014:36616:36771 [0] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported
node014:36617:36769 [1] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported
node014:36617:36769 [1] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported
node014:36617:36769 [1] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported

What do these message mean? Will they impact the performance?

**Comment by alsrgv** [comment]
@hfutxrg, could you try running on synthetic data first, without --data_dir & --data_name?

`node014:36616:36771 [0] INFO nvmlDeviceGetNvLinkRemotePciInfo() failed: Not Supported` can be ignored - they're just stating that your system doesn't have NVLink.

We found `-mca btl_openib_receive_queues P,128,32:P,2048,32:P,12288,32:P,131072,32` to be helpful, can you clarify what do you mean by "does not work for me" ?

Can you also do mpirun with 4 GPUs in a single node, what speedup do you get?

**Comment by hfutxrg** [comment]
OK, I will try with synthetic data first and also run 4 GPUs on single node.

Because I only tried 2 nodes and therefore I used ```-mca btl_openib_receive_queues P,128,32:P,2048,32```, but then the program could not continue after the running warm up and outputting the devices info. It seems the program was hanging.

**Comment by alsrgv** [comment]
Let me know how it goes!  To isolate issue you can also try `-x NCCL_IB_DISABLE=1 -mca btl ^openib` to disable IB altogether.

**Comment by hfutxrg** [comment]
@alsrgv sure. I will let you know when I get the performance numbers.

**Comment by hfutxrg** [comment]
@alsrgv I have finished the testing. I found with synthetic data, my speedup is 1.97x, and with real imagenet data, the speedup is 1.45x. How does Horovod read the input data? Using single thread or multiple threads? Or the work of processing input data is done by TensorFlow?

**Comment by alsrgv** [comment]
@hfutxrg Is 1.97x performance of 1 machine vs 2 machines, or 8 gpus vs 1 gpu?

The real data is handled completely by TensorFlow.  One suspicion I have is whether CPU are powerful enough to handle multiple streams of data.  Can you try couple of things with real data:

1. `mpirun -np 1 <real data training>`
2. `git checkout horovod; mpirun -np 4 <real data training>` on 1 machine
3. `git checkout master; mpirun -np 4 <real data training>` on 1 machine <-- this last one will essentially just run 4 training processes without communication with each other, this can help see if CPU is able to catch up with all the preprocessing
4. Try (2) and (3) with `--distortions False` which disables CPU preprocessing, but will hurt accuracy.

**Comment by hfutxrg** [comment]
@alsrgv both 1.97x and 1.45x are the speedup of using 2 nodes over 1 node, with 4 GPUs on each node. So it's actually speedup of 8 GPUs vs 4 GPUs. 

Sure I will try those different combinations and let you know the results.

**Comment by alsrgv** [comment]
Gotcha. I updated instructions above to make it a little simpler - no need to switch branches.

**Comment by hfutxrg** [comment]
@alsrgv I used a node with 4 Volta GPUs (V100). The following are the results for different combinations (all of them use real data training):
1. np=1, speed is **321.7** images/sec
2. np=4, speed is **917.2** images/sec, 2.85x over np=1
3. np=4, allreduce=false, speed is **1214.1** images/sec, 3.77x over np=1
4. np=4, distortions=false, speed is **1150.2** images/sec, 3.58x over np=1
5. np=4, allreduce=false, distortions=false, speedup is **1256.6** images/sec, 3.91x over np=1

**Comment by alsrgv** [comment]
Since speedup improves a lot when distortions are disabled, CPU may be playing a role here.  What may be happening is that it can't handle doing 4x more distortion work in synchronous mode (2), but can spread the work in test (3).  Do you know what kind of CPU do you have?  You can check by running `dmesg | grep CPU`.

Another thing to try is to shift preprocessing from CPU to GPU by adding `with tf.device('/gpu:0'):` to https://github.com/alsrgv/benchmarks/blob/horovod/scripts/tf_cnn_benchmarks/preprocessing.py#L346

**Comment by hfutxrg** [comment]
Hi @alsrgv, the CPU on my server is Intel Xeon E5-2690 v4 @2.6GHz. I think it's powerful enough.

I tried to use ```with tf.device('/gpu:0')``` for ```distort_image()```,  but the training speed is lowered to 804.4 images/sec on single node.

I also tried ```opt = hvd.DistributedOptimizer(opt, device_dense='/cpu:0')```, then the training speed is lowered 865.8 images/sec on single node. 

Is there any other optimization that can be done without disabling distortions?

**Comment by alsrgv** [comment]
@hfutxrg very interesting data points!  I will run a real data experiment next week on our server with Voltas and let you know if I see a similar or different scaling.

**Comment by hfutxrg** [comment]
@alsrgv I also tried two nodes where each node has 4 P100 GPUs, and the speedup with 2 nodes is ~1.59x over one node. This speedup is a little bit higher than V100 nodes but still far from the results on Horovod readme. So the speedup on Horovod readme file is for synthetic data rather than real data, right?

**Comment by alsrgv** [comment]
@hfutxrg, correct - the idea was to separate GPU workload that involves synchronization from preprocessing which in theory is easily parallelizable. I didn't have a chance to run real data tests yet, once I do I'll let you know if I'm seeing similar phenomena.

**Comment by alsrgv** [comment]
@hfutxrg I got some results on a server with 4x**P100** GPUs for starters:

batch_size=64: **3.3x** with resnet50, and **3.5x** with resnet101.
batch_size=128: **3.7x** with resnet50.

You can also experiment with `config.inter_op_parallelism_threads` option. I had slightly better performance with the value of `8`:

batch_size=64: **3.5x** with resnet50, and **3.6x** with resnet101.
batch_size=128: **3.7x** with resnet50.

**Comment by hfutxrg** [comment]
Hi @alsrgv, could you send your outputs of running resnet50 with 1 P100 and 4 P100? I also used batch_size=64 but only got 2.85x for 4 P100. I want to see what is the difference between your output and my output. Yes I will also try to use a larger value of ```config.inter_op_parallelism_threads```.

**Comment by alsrgv** [comment]
@hfutxrg:

* `config.inter_op_parallelism_threads = 8`: 213 img/sec on 1xP100, 758 img/sec on 4xP100 - 3.55x speedup.
* `config.inter_op_parallelism_threads = unlimited`: 217 img/sec on 1xP100, 719 img/sec on 4xP100 - 3.31x speedup.

Note that this is SXM2 (not PCI-e) server.  How many DIMMs do you have on the server?  We have 4x32GB DIMMs on our server balanced across memory channels.  More channels the better for training.

**Comment by hfutxrg** [comment]
@alsrgv I found in my last post, there was a mistake. The 2.85x speedup is actually for 4 V100-PCIe. 
I also tried P100-SXM2 with batch_size=64, and the speedup is 3.16x. The performance is 233 images/sec on 1xP100, 736 images/sec on 4xP100. 

All of my servers have 16x16GB DIMMs. So here it seems faster GPUs lead to lower speedup. I will also try larger ```config.inter_op_parallelism_threads``` to see whether there is any improvement.

**Comment by alsrgv** [comment]
@hfutxrg I just tried AWS V100 server. On 4xV100 I get 1179 img/sec. When I try a single GPU, some silliness is happening and the performance is just 77 img/sec. What's your 4xV100 number?

**Comment by hfutxrg** [comment]
My number for 4 V100-PCIe is only 917 images/sec. Your 1179 images/sec is for default configuration or setting ```config.inter_op_parallelism_threads = 8 ``` or ```unlimited```? 

Also on your V100 server, what is the CPU model, memory size, V100-PCIe or V100-SXM2, and how GPUs are connected to CPUs (use "nvidia-smi topo -m" command)?

**Comment by alsrgv** [comment]
@hfutxrg, unfortunately, I cannot share the spec at the moment. Did you have a chance to try out `config.inter_op_parallelism_threads = 8`?

**Comment by hfutxrg** [comment]
Hi @alsrgv , in one V100-SXM2 node where two CPUs are connected to four GPUs with two PCIe buses, ```config.inter_op_parallelism_threads = 7``` improved ~25 images/sec. I chose 7 instead of 8 because my node has 28 cores so that each GPU can run on 7 cores. But from ```top``` command, I still saw each GPU only used 5 CPU cores. I checked that the default value ```inter_op_parallelism_threads``` in TensorFlow is 5. So I don't know why each GPU still used 5 cores instead of 7 cores, although the performance improved a bit. I also tried to set it to 8 but the performance is 16 images/sec lower than setting to 7.

But this setting does not improve the performance in my other two nodes. In one node, two CPUs are connected to four GPUs by only one PCIe bus, and the GPUs are V100-SXM so that they have P2P accesses. In another node, two CPUs are connected to four GPUs by two PCIe buses, and the GPUs are V100-PCIe and they do not have full P2P accesses. 

I did the multi-node testing with the above two types of nodes. With V100-PCIe nodes, the scaling stopped from 4 nodes, and with V100-SXM2 nodes, the scaling stopped from 3 nodes. I checked with ```nvidia-smi``` tool and found lots of time the GPUs utilization are 0% in those nodes. 

So will the mode performance in your MPI implementation or TensorFlow implementation be impacted by the number of PCIe buses between CPU and GPUs, and the P2P accesses between GPUs. Do you have any suggestions on how to improve the multi-node performance in my cases? Thanks.

**Comment by alsrgv** [comment]
@hfutxrg, can you collect [Horovod Timeline](https://github.com/uber/horovod/blob/master/docs/timeline.md) for your training?  That may help reveal areas to dig in further.  Also, I assume that your training data is on the local disk with fast access speed?

**Comment by hfutxrg** [comment]
@alsrgv Thanks for the timeline tool suggestion. I will give it a try. Right now the training data is stored on a HDD NFS. I will try to copy it to the local SSD in each node to see whether there is any improvement.

**Comment by alsrgv** [comment]
@hfutxrg, aha! NFS latency may be your bottleneck in this case. Let's see results on SSD.

**Comment by alsrgv** [comment]
@hfutxrg, did reading data from local SSD improve your scaling?

**Comment by hfutxrg** [comment]
Hi @alsrgv, using SSD has the same performance as using NFS. However, I found there are many "waiting for data" in the Horovod timeline, so I suspect reading the dataset is the bottleneck. Then I found TensorFlow website says "Placing input pipeline operations on the CPU can significantly improve performance". So i tried to use the following code:
```with tf.device('/cpu:0'):```
 ```           image = distort_image(image, self.height, self.width, bbox, thread_id)```

You suggested me to use ```with tf.device('/gpu:0')``` but it seems using cpu is better. Right now the speedup of using 16 V100(4 nodes) is 9.39x compared to 1 V100. 

This is an acceptable speedup, but I want to see whether it can be further improved. I found the code that reads the input dataset is as follows:
```record_input = data_flow_ops.RecordInput(```
    ```      file_pattern=dataset.tf_record_pattern(subset),```
        ```  seed=301,```
      ```    parallelism=64,```
       ```   buffer_size=10000,```
        ```  batch_size=self.batch_size,```
         ``` name='record_input')```

Do you think whether the performance can be improved by tuning "parallelism" and "buffer_size"? I don't know what the exact value should be for these parameters, but I will definitely try some.

**Comment by alsrgv** [comment]
@hfutxrg, it may be interesting to tinker with these parameters. If you're using same seed everywhere, you may actually be reading the same files at the same time, which may be hurting FS performance. You can try to specify different `seed` instead, e.g. `seed=hvd.rank()`. Parallelism may also benefit from being reduced, since you're running 4 processes on one server, and so it will try to run 256 total threads which are far more than the number of cores that you have.

**Comment by hfutxrg** [comment]
@alsrgv, I tried to reduce the parallelism and change ```seed=hvd.rank()```, but none of them can improve the performance. However, I tried the option "-mca btl_openib_receive_queues P,128,32:P,2048,32:P,12288,32:P,131072,32" again for 4 nodes and it did improve the performance. Now the speedup of using 16 GPUs (4 nodes) is 12.45x compared to 1 GPU which is pretty good. What I do not understand is how did you come up with these numbers ```P,128,32:P,2048,32:P,12288,32:P,131072,32```. I searched online and found that "P" means "Per-peer receive queues", and the two numbers following it represent the message size and num_buffers. But why did you choose these numbers? If I have more nodes, how do I choose these values?

Another question: I saw another branch horovod_v2 and it has fp16 support. Is it ready to test? I think it also requires TensorFlow to support FP16, right? But it seems even the latest TensorFlow still does not support FP16, is that correct?

**Comment by robzor92** [comment]
Hello @alsrgv, I've been performing the benchmarks for the SingleRoot supermicro machines on behalf of @jimdowling 

The first numbers we publish is for a single machine (10x1080 GTX Ti)
All the runs were with the following command using synthetic datasets.

mpirun -np **num_gpus** python ./scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py --model **model_name** --batch_size 64 --horovod_device gpu

The scalability we've observed corresponds to the trend in the following picture
![resnet101_benchmark](https://user-images.githubusercontent.com/9936580/32720351-9e15037a-c863-11e7-9bd2-3b7cb2d2f40b.png)

As you can see we achieve near perfect scaling up to 5 GPUs. Beyond that we see the impact of having the ring formed over two different buses, but still very impressive.

**ResNet-50**
1 GPU: 188.35 images/sec
10 GPUs: 1685 images/sec

**ResNet-101**
1 GPU: 115.53 images/sec
10 GPUs: 1035.03 images/sec

**VGG-16**
1 GPU: 126.8 images/sec
10 GPUs: 965 images/sec

**Inception V3**
1 GPU: 132.8 images/sec
10 GPUs: 1170.25

Also, we had to run the benchmarks on the _horovod_ branch of the benchmarks repo.
The _horovod_v2_ branch printed output like this

> 2017-11-13 11:27:49.810770: E tensorflow/stream_executor/cuda/cuda_driver.cc:936] failed to allocate 224.38M (235274240 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY
2017-11-13 11:27:49.811557: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 9 and 0
2017-11-13 11:27:49.816120: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 9 and 8
2017-11-13 11:27:49.816146: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1045] Device peer to peer matrix
2017-11-13 11:27:49.818031: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 9 and 1
2017-11-13 11:27:49.824038: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 9 and 2
2017-11-13 11:27:49.828037: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 8 and 9
2017-11-13 11:27:49.829379: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 9 and 3
2017-11-13 11:27:49.835285: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 9 and 0
2017-11-13 11:27:49.835573: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 9 and 4
2017-11-13 11:27:49.839939: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 9 and 1
2017-11-13 11:27:49.840242: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 9 and 5
2017-11-13 11:27:49.852928: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 9 and 2
2017-11-13 11:27:49.853208: W tensorflow/core/common_runtime/gpu/gpu_device.cc:918] Unable to enable peer access between device ordinals 9 and 6
2017-11-13 11:27:49.854527: E tensorflow/stream_executor/cuda/cuda_driver.cc:936] failed to allocate 224.38M (235274240 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY

In the next benchmarks we will publish benchmark for running multiple machines hooked up with RDMA.

Finally I have some questions for making the benchmarks more interesting.

Do you have any pointers for running the benchmarks with real testing datasets?
What command/branch was used to run the Distributed TensorFlow benchmarks?

**Comment by alsrgv** [comment]
@robzor92, thanks for sharing your results! I think these look good for 10xGPU machine, considering how much gradient exchanges put on PCIx bus.

When you try `horovod_v2` branch, can you add `--variable_update horovod` argument?  You can see detailed instructions [here](https://github.com/uber/horovod/blob/master/docs/benchmarks.md).

We used the master branch from https://github.com/tensorflow/benchmarks at the time of writing. The command used was: `python tf_cnn_benchmarks.py --model {network} --batch_size 64 --num_gpus 4 --worker_hosts {worker_hosts} --ps_hosts {ps_hosts} --task_index {task_index} --job_name {job_name} --local_parameter_device cpu`. Each host had one worker and one parameter server, and placeholders were substituted appropriately.

**Comment by jimdowling** [comment]
If anybody is interested in how we run Horovod on Spark, to reproduce @robzor92 experiments, i gave a talk here:
https://www.youtube.com/watch?v=5bVubfUt7ec
We will be releasing Spark/Horovod with the Hops platform in the next few weeks.

**Comment by zh794390558** [comment]
@robzor92 The result looks good.

**Comment by RookieoftheYear** [comment]
> @alsrgv, I tried to reduce the parallelism and change `seed=hvd.rank()`, but none of them can improve the performance. However, I tried the option "-mca btl_openib_receive_queues P,128,32:P,2048,32:P,12288,32:P,131072,32" again for 4 nodes and it did improve the performance. Now the speedup of using 16 GPUs (4 nodes) is 12.45x compared to 1 GPU which is pretty good. What I do not understand is how did you come up with these numbers `P,128,32:P,2048,32:P,12288,32:P,131072,32`. I searched online and found that "P" means "Per-peer receive queues", and the two numbers following it represent the message size and num_buffers. But why did you choose these numbers? If I have more nodes, how do I choose these values?
> 
> Another question: I saw another branch horovod_v2 and it has fp16 support. Is it ready to test? I think it also requires TensorFlow to support FP16, right? But it seems even the latest TensorFlow still does not support FP16, is that correct?

hi，could you please tell me how to use the option "-mca btl_openib_receive_queue" in horovodrun command? thx!

**Comment by tgaddair** [comment]
Hey @RookieoftheYear, you can set this using `--mpi-args "-mca btl_openib_receive_queue"`

**Comment by RookieoftheYear** [comment]
> Hey @RookieoftheYear, you can set this using `--mpi-args "-mca btl_openib_receive_queue"`

thansk for your reply , but the option did not work with mine , i get same problem with [this](https://github.com/horovod/horovod/issues/476#issue-356341320)，I have been tortured by this problem for weeks

**Comment by tgaddair** [comment]
Hey @RookieoftheYear, to clarify, you're seeing long QUEUE times in the timeline?  Do you also see this behavior when you run the synthetic benchmark scripts found in the examples section?  It might make sense to create a new issue so we can dig into the details of your environment and use case.

**Comment by RookieoftheYear** [comment]
* 1 node with 2 gpus（two GeForce RTX 2070）：
  * command:  `horovodrun -np 2 --mpi-args "-mca btl_openib_receive_queues P,128,32:P,2048,32:P,12288,32:P,131072,32"  --timeline-filename ./result/timeline-node1-2gpus.json -H localhost:2 python tensorflow2_synthetic_benchmark.py --batch-size=64 |& grep -v "Read -1"`
  * result:  `Total img/sec on 2 GPU(s): 342.5 +-2.1`
  * timeline
![node1-2gpus](https://user-images.githubusercontent.com/20176796/99622996-258d9a80-2a66-11eb-9edd-9993b2d519ba.png)

* 1 node( the remote server) with 1 gpu (GeForce GTX 1080)

  * command:  `horovodrun -np 1 --mpi-args "-mca btl_openib_receive_queues P,128,32:P,2048,32:P,12288,32:P,131072,32"  --timeline-filename ./result/timeline.json -H localhost:1 python tensorflow2_synthetic_benchmark.py --batch-size=64 |& grep -v "Read -1"`

  * result:  `Total img/sec on 1 GPU(s): 137.5 +-0.7`

* 2 nodes with 3 gpus

  * command:  `horovodrun -np 3 --mpi-args "-mca btl_openib_receive_queues P,128,32:P,2048,32:P,12288,32:P,131072,32"  --timeline-filename ./result/timeline-2nodes-3gpus.json -H localhost:2,dp2:1 -p 12345 python tensorflow2_synthetic_benchmark.py --batch-size=64 |& grep -v "Read -1"`

  * result： `Total img/sec on 3 GPU(s): 211.7 +-14.0`

  * timeline
![node1 2-3gpus](https://user-images.githubusercontent.com/20176796/99622946-13abf780-2a66-11eb-9b8d-669ccd2f0994.png)
